src/K.hs

```diff
@@ -157,16 +162,21 @@ kExprBool :: Exp Bool -> String
 kExprBool (And a b) = "(" <> kExprBool a <> " andBool\n " <> kExprBool b <> ")"
 kExprBool (Or a b) = "(" <> kExprBool a <> " orBool " <> kExprBool b <> ")"
 kExprBool (Impl a b) = "(" <> kExprBool a <> " impliesBool " <> kExprBool b <> ")"
-kExprBool (Eq a b) = "(" <> kExprInt a <> " ==Int " <> kExprInt b <> ")"
-
-kExprBool (NEq a b) = "(" <> kExprInt a <> " =/=Int " <> kExprInt b <> ")"
 kExprBool (Neg a) = "notBool (" <> kExprBool a <> ")"
 kExprBool (LE a b) = "(" <> kExprInt a <> " <Int " <> kExprInt b <> ")"
 kExprBool (LEQ a b) = "(" <> kExprInt a <> " <=Int " <> kExprInt b <> ")"
 kExprBool (GE a b) = "(" <> kExprInt a <> " >Int " <> kExprInt b <> ")"
 kExprBool (GEQ a b) = "(" <> kExprInt a <> " >=Int " <> kExprInt b <> ")"
 kExprBool (LitBool a) = show a
 kExprBool (BoolVar a) = kVar a
+kExprBool (NEq a b) = "notBool (" <> kExprBool (Eq a b) <> ")"
+kExprBool (Eq (a :: Exp t) (b :: Exp t)) = case eqT @t @Integer of
+  Just Refl -> "(" <> kExprInt a <> " ==Int " <> kExprInt b <> ")"
+  Nothing -> case eqT @t @Bool of
+    Just Refl -> "(" <> kExprBool a <> " ==Bool " <> kExprBool b <> ")"
+    Nothing -> case eqT @t @ByteString of
+      Just Refl -> "(" <> kExprBytes a <> " ==K " <> kExprBytes b <> ")" -- TODO: Is ==K correct?
+      Nothing -> error "Internal Error: invalid expression type"
```

Comment:
@MrChico this is a little clunky, is there some nice trick with the Maybe monad I can use to tidy this up a bit?

Reply:
hmm, the fact that you are proceeding in the `Nothing` case makes it a bit tricky... I don't immediately see a cleaner way